### ISSUE TITLE
Colon in string with a space after it can lead to invalid param name

### ISSUE THREAD

**Issue by MaerF0x0** [jmoiron/sqlx#206]
```SQL
SELECT '1: profiles' as p FROM profiles
```

Will result in an empty string name being inserted into the param name list.

This line should only run if we find a non-empty name: https://github.com/jmoiron/sqlx/blob/master/named.go#L264-L265

We're seeing errors:

```golang
"could not find name  in map[string]interface {}(nil)"
```
note the dbl space between "name" and "in" , its substituting an empty string in %s.  

LMK if you'll take a PR for fixing.

**Comment by MaerF0x0** [comment]
@jmoiron  Ping.   Do you agree this is a real bug?

**Comment by flibustenet** [comment]
Maybe you should use args ("select $1 as p from profiles", "1: profiles") ?
I think there are some places in the code where we can see that quotes are not excluded from parsing. Don't know if it's to patch or if it's to force the use of params ?

**Comment by jmoiron** [comment]
There was another issue in the past with named queries and not being able to escape colons.  There is another issue which is also essentially a duplicate, 
In a way this is a duplicate of https://github.com/jmoiron/sqlx/issues/166.  I think my stance currently is:

* sqlx should stay out of the game of trying to be smart with parsing
* queries that use some extended sqlx features (like named queries) should not be subtly invalid to run via other means
* I prefer the named interface to have obvious, straightforward limitations than complex nuanced ones, even if the obvious ones are more onerous
* named queries are an optional nice-to-have and if you need to use colons you can fall back to non-named queries


The first and third rules rule out one attempt, which was trying to parse where `:` was used inside strings.  It's used elsewhere in various dialects, eg in pg type casting, and it's a losing game to try to get a handle on that.

Another option would be to use and detect `::` as a literal colon.  This is, I believe, how `'` is escaped in the SQL standard (this may not be correct, but it may be de-facto correct for modern implementations).  This violates 2 in a somewhat subtle manner, and it would require queries to have `::::` when double-colons (present, again, in pg type exprs) are necessary, which looks wrong.

In the end, I'm not _really_ happy with any of the "solutions" to this problem.  Even @flibustenet's proposal won't work because you can use colons in places where you cannot supply arguments, and in fact that is the case in the example.

I'm open to suggestion.

**Comment by MaerF0x0** [comment]
I see how this is quite difficult due to being inside a single quotes.  I figured it was a simple issue because 

1.  the colon is prepended by nonwhitespace chars ("1") 
1.  the colon only has whitespace after it and thus cannot be a valid substitution

so I would propose to at least implement rule 2 (ignore colons followed by whitespace)

**Comment by detaoin** [comment]
I might have a solution, however it goes against the _sqlx shouldn't be smart_ rule: a simplified SQL tokenizer. I already have a single file implementation which seems to work for me.

I put the code together in PR #474 so you can have a look. The idea is to catch named parameters only if they correspond to a valid SQL identifier (not in quotes or comments).
However the problem is the performance drop.

I don't know if this is an acceptable solution to the issue.

**Comment by detaoin** [comment]
Yes I think that converting this tokenizer into its own project might be the best approach. I'm using it anyway for another project of mine ([sql2http](https://github.com/detaoin/sql2http)). So I can work on better testing and performance improvements.

I will do this (however I don't have much time right now, so the progress might be a bit slow), and keep you informed when there are news.

Thanks for the feedback.

**Comment by firu11** [comment]
I've discovered this as well. In a comment though.

```go
r, err := sqlx.NamedQueryContext(ctx, txx, `
insert into chat (
  chat_name,
  share_key, -- TODO: something
  user_id,
) values (
  :chat_name,
  :share_key, -- TODO: something
  :user_id
) returning chat_id`, chat)
```

`could not find name  in map[string]interface {}(nil)`
                                      ^
